---
title: "assembly_analysis"
author: "Ram Krishna Shrestha"
date: "31 January 2019"
output: html_document
---


## Simulation of Illumina Short Read Dataset

A tool - ART was used for simulation of Illumina short read datasets from randomly generated Renseq Enriched DNA fragments in FASTA format. In order to observe the impact of different variables in the simulated datasets while reads assembly, short reads were simulated with different sets of parameters. Threre is difference of one parameter in each simulated dataset. The parameters varied in each data simulation are: 
1) Illumina Platform Model (Hiseq, MiSeq, NextSeq)
2) Short read length (HiSeq - 125, 150; MiSeq - 250, NextSeq- 75)
3) Coverage of the data generated (10, 20, 30 ,40, 50)
4) Meansize of fragments (200, 300, 400, 500)
5) Standard deviation of fragments (15, 20, 25)
6) Minimum Quality of simulated reads (10,20,30,40)
7) Max insertions in the simulated reads (1,2,3,4,5)

## Assembly of simulated short read dataset

Simulated short read datasets are assembled using ABYSS tool. We will also use soapdenovo in the future. I would also like to observe the impact of assembly tool input parameters in the quality metrics of the final assembly. I choosed the following input parameters for assembly:

1) Kmer (21,25, 29, 33, 37, 41, 45, 49, 53, 55)
2) Min coverage (3, 5, 10)
3) Min kmer coverage (10, 20, 30, 40, 50)

## Assembly Results

The contigs shorter than 200 bp are filtered out. The analysis is done on the contigs that passed the length threshold.

As we have many parameters, we will fix some parameters value and see the effect of the rest of the parameters on the assembly metrics.

### Effect of Kmer, MinCoverage and MinKmerCoverage on the number of contigs in the assemblies


```{r, echo=TRUE}
library(ggplot2)
library(dplyr)
# read the table
df=read.table('~/assembly_comparison/results/assembly_analysis_report.csv', sep=",", header=TRUE)
```

```{r, echo=TRUE, fig.width=12, fig.height=8}
ggplot(df, aes(x=as.factor(Kmer), y=Contigs)) + geom_point(aes(color=as.factor(MinKmerCoverage))) + facet_wrap(~as.factor(MinCoverage))
```

The plot above shows that at low kmer coverage the final number of contigs in the assembly is high. However this is not desirable as this could mean a lot of contigs are short in length or highly fragmented assembly. The downstream process can be difficult with very short assembly contigs. Choosing higher Kmer length, reduces the number of contigs significantly than lower Kmer length.

### Effect of Kmer, MinCoverage and MinKmerCoverage on the maximum contig length in the assemblies

```{r, echo =TRUE, fig.width=12, fig.height=8}
ggplot(df, aes(x=as.factor(Kmer), y=MaxContigLength)) + geom_point(aes(color=as.factor(MinKmerCoverage)), position=position_dodge(width=0.5) )  + geom_boxplot() + facet_wrap(~as.factor(MinCoverage))
```

The Max contig length is higher at lower min kmer coverage and higher kmer length. This means that fewer kmers were enough for elongating the contigs and this could be errorneous. There is significant difference in max contig length at lower kmer coverage of 10 and 50.

### Effect of Bait mapped ratio between assembly and refseq

```{r, echo = TRUE}
ggplot(df, aes(RatioBaitToRef)) + geom_density(aes(colour=as.factor(Kmer), fill=as.factor(Kmer)))
```
The above plot shows the ration of baits between assembly and the refseq.

```{r, echo = TRUE}
ggplot(df, aes(RationBaitToRandSeq)) + geom_density(aes(colour=as.factor(Kmer), fill=as.factor(Kmer)))
```

The above plot shows the ratio of baits between assembly and random simulated sequence reads.
